lib/attrsets.nix

```diff
@@ -74,10 +74,34 @@ rec {
        => error: cannot find attribute `z.z'
   */
   getAttrFromPath = attrPath: set:
-    let errorMsg = "cannot find attribute `" + concatStringsSep "." attrPath + "'";
+    let errorMsg = "cannot find attribute `${concatStringsSep "." attrPath}`";
     in attrByPath attrPath (abort errorMsg) set;
 
 
+  /* Apply a function to an attribute in a nested attribute set. If it doesn't
+     find the path it will throw.
+
+     Example:
+       mapAttrByPath ["a"] { a = { b = 5; }; } (x: { c = x.b - 2; })
+       => {a = { c = 3; }; }
+  */
+  mapAttrByPath = attrPath: set: f:
+    let
+      errorMsg = "cannot find attribute `${concatStringsSep "." attrPath}`";
+      recur = attrPath: set:
+        if attrPath == [] then
+          f set
+        else
+          if set ? ${head attrPath} then
+            set // {
+              ${head attrPath} = recur (tail attrPath) set.${head attrPath};
+            }
+          else
+            abort errorMsg;
```

Comment:
"map" functions generally don't throw, but return a value equal to the original value when the substructure is missing. E.g. `map throw [] == []`, `mapAttrs throw {} = {}`, so I would expect `mapAttrByPath ["x" throw] {} throw == {}`.

Reply:
You can take advantage of `attrPath` in the error message, to show the missing/wrongly typed attribute name and to compute the path that lead up to it (using the original `attrPath`; need to un-shadow the original `attrPath` argument).

This won't cost performance because everything you need is already in scope if you fix the shadowing.